### ISSUE TITLE
SameDiff: (snapshots) imported BERT model producing NaNs (CPU)

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/deeplearning4j#8512]
When running BERTGraphTest.testBert, I get NaNs in one of two possible locations:

* softmax - bert/encoder/layer_3/attention/self/Softmax
* add - bert/encoder/layer_2/output/LayerNorm/batchnorm/add_1

In both cases I have confirmed there are no NaNs in the input.

This appears to be related to SameDiff's array reuse memory management strategy, ArrayCacheMemoryMgr.
Using the simpler (but lower performance) ArrayCloseMemoryMgr (INDArray.close() after arrays are no longer needed) and NoOpMemoryMgr (clear references + rely on GC) the test looks ok, no NaNs.

Adding this just before the SameDiff.output call results in correct results:
```
InferenceSession is = new InferenceSession(sd);
is.setMmgr(new ArrayCloseMemoryMgr());
//is.setMmgr(new NoOpMemoryMgr());    //Results also ok with this
sd.getSessions().put(Thread.currentThread().getId(), is);
```

Still working to reproduce this in isolation. I don't believe it's a problem with the strategy per-se (incorrectly re-using arrays when they shouldn't be), given the reproducible and isolated nature of the bug (plus, the fact that we use the same array reuse strategy without issues in literally every other SameDiff test, other than DL4J SameDiff layers).

**Comment by SidneyLann** [comment]
should be the same reason as #8496

**Comment by AlexDBlack** [comment]
Looks like the underlying issue is a (recently introduced?) problem with broadcast mul op for some combinations of input shapes.

**Comment by AlexDBlack** [comment]
Confirmed fixed and BERT import tests passing for both CPU and CUDA.
Will be merged back to eclipse/deeplearning4j soon (and hence in snapshots soon) and also in the upcoming release.
